### DISCUSSION TITLE
Using client_credentials flow for machine-to-machine communication

### DISCUSSION THREAD

**Discussion by aovasylenko** [zitadel/zitadel#4009] [Q&A] [answered]
Some Auth provider proposing to use client_credentials flow for machine-to-machine communication, which means registering apps as OAuth2 clients, which sounds a bit as a hack. 
I like the concept of service users, where they are represented with a user_id. However the process of token issuing (signing json manually with a private key) could be a bit complicated for 3rd parties of out APIs (where we want to issue them a token to access our APIs). Is there another way to do it? For example have a 'password/service' for service users and user `password` grant flow to issue the token for service account user?

**Comment by mffap** [comment]
You can also use Private Access Tokens that you can create for Service Users. Would that be an option?
A detailed guide is missing: https://github.com/zitadel/zitadel/issues/3678

**Comment by aovasylenko** [comment, marked as answer]
@mffap this is super helpful! I'm not sure what is the right way to extract information about user (validate token)? I tried to use:
```
curl --request POST   --url https://<DOMAIN>/oidc/v1/userinfo   --header 'Content-Type: application/x-www-form-urlencoded'   --header 'Authorization: Bearer <TOKEN>'
```
but getting 
```
{"Parent":null,"Message":"token is not valid or has expired","ID":"OIDC-Dsfb2"}
```

**Reply by fforootd** [reply]
Hi @aovasylenko 

Maybe I can help here 👋 

Since the current PAT is only available to service users it is not intended to be used on the `userinfo endpoint`. The `userinfo endpoint` is generally seen as API for GUIs with an interactive user.

However you can check the PAT against the[introspect endpoint](https://docs.zitadel.com/docs/apis/openidoauth/endpoints#introspection_endpoint) if you setup an API client in your project.

@livio-a any opinion from you on this?

**Reply by aovasylenko** [reply]
Thank you @fforootd !
So, yes, I have progress:
1. I've created Machine-API and got client_id and client_secret
2. I've created Service user and got token
3. I've used example code to call introspect API but received: `{"active":false}`. Wondering if I need authorize this service user in some way?

Also I noticed, that API response takes 1 second - is it intended ?

**Reply by fforootd** [reply]
> So, yes, I have progress:
> 
> I've created Machine-API and got client_id and client_secret
> I've created Service user and got token
> I've used example code to call introspect API but received: {"active":false}. Wondering if I need authorize this service user in some way?

Which example code are you using? Just so that I can have a look into this 😄 

> Also I noticed, that API response takes 1 second - is it intended ?

That's a thing we are currently discussing internally (@hifabienne @livio-a . The current settings for the password hashing are to slow. Self-hoster can change [this with this setting](https://github.com/zitadel/zitadel/blob/v2-alpha/cmd/defaults.yaml#L263). 
We provide the best latency when you use [JWT Profile for client authentication](https://docs.zitadel.com/docs/apis/openidoauth/grant-types#using-jwts-for-client-authentication) since this uses RSA256 it is much faster.

May I ask what language you are using. Because some support this out of the box or we provided a SDK 😁

**Reply by fforootd** [reply]
Hi @aovasylenko 

I quickly wanted to come back to you again.
We found a bug in the PAT implementation #4024 which prevented usage of the PAT tokens altogether.

This should be soon included in a new release.

**Reply by livio-a** [reply]
Hey @aovasylenko 

The fix was released in version [v2-alpha.44](https://github.com/zitadel/zitadel/releases/tag/v2.0.0-v2-alpha.44). We will deploy this version onto zitadel.cloud shortly.

**Reply by livio-a** [reply]
Hey @aovasylenko 

We've just released [version 2.19](https://github.com/zitadel/zitadel/releases/tag/v2.19.0) with support for client_credentials on service users. If you run you own ZITADEL instance, you can already try it out: https://zitadel.com/docs/guides/integrate/client-credentials
We will deploy it to zitadel.cloud on Monday evening.
